### ISSUE TITLE
Modularize rules

### ISSUE THREAD

**Issue by olafurpg** [scalacenter/scalafix#818]
Currently, scalafix has accumulated a decent number of rules. Some of those rules would belong better in a separate repo so that they can evolve at a different pace than scalafix.

- `Disable`: a lot of complexity and configuration options. I'd like to keep scalafix stable after v0.6 and given the number of requests to improve/tweak this rule I believe it's best to have it in a separate repo.
- `MissingFinal`: does a combination of fixing and linting
- `Dotty*`: can evolve better as a separate project the same way as the collection compat rules.
- `ExplicitResultTypes`: the implementation is low quality and there exist better alternatives like [`RscCompat`](https://github.com/twitter/rsc/blob/5815b87401bc949526f580dfe71cf017cf6fd990/scalafix/rules/src/main/scala/rsc/rules/RscCompat.scala). There's a big design space to consider when it comes to this problem so I think it makes more sense to keep it outside of the core repo.

This leaves us with only a handful of rules in the core repo. To begin with, this may hurt since there's less functionality out-of-the-box. However, in the long run I believe it's the best strategy for keeping Scalafix maintainable as a project.

**Comment by olafurpg** [comment]
I agree, the current `core` module contains far too much. Service loaders are working so great that I'm tempted to remove the hardcoded list of builtin rules and list builtin rules in `META-INF/services/scalafix.v1.Rule` instead. The `rules` module would then not get any special treatment besides being a dependency of `scalafix-cli` :)

**Comment by gabro** [comment]
Ah yes, that would be great :)

**Comment by olafurpg** [comment]
I want to stress that removing rules like `ExplicitResultTypes` does not mean they will require custom installation if they are maintained in a separate repo. I am open to include high-quality rule implementations by default by adding them as dependencies to sbt-scalafix. The primary motivation is to

- enable rules to evolve at a faster pace than Scalafix core. I think the best path towards having high quality linters and rewrites is to enable more people to create rules that they want to use themselves and have ownership over their development. Having dozens of "blessed" but semi-unmaintained rules in the core repo gives a false sense of security and is likely to hinder innovation.
- allow Scalafix itself to stabilize as a project. The issue tracker will be reserved for maintaining core functionality like `Patch` and `SemanticDoc`.

**Comment by olafurpg** [comment]
If there is demand, I am even open to document high-quality custom rules on the scalafix website. The docs for those rules would need to make it clear

- who maintains the rule
- where to report issues
